### DISCUSSION TITLE
Enabling alpha feature gate doesn't work

### DISCUSSION THREAD

**Discussion by jayzziebone** [k3s-io/k3s#6580] [Q&A]
**Environmental Info:**
K3s Version: v1.22.16+k3s1

Node(s) CPU architecture, OS, and Version: 

Linux node-10-120-13-171 5.4.0-70-generic #78~18.04.1-Ubuntu SMP Sat Mar 20 14:10:07 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

Cluster Configuration:

NAME                  STATUS   ROLES                  AGE   VERSION
node-10-120-13-171    Ready    control-plane,master   12h   v1.22.16+k3s1
node-10-120-127-171   Ready    <none>                 12h   v1.22.16+k3s1
node-10-120-127-172   Ready    <none>                 12h   v1.22.16+k3s1
node-10-120-127-173   Ready    <none>                 12h   v1.22.16+k3s1
node-10-120-127-170   Ready    <none>                 12h   v1.22.16+k3s1

**Describe the bug:**

**Steps To Reproduce:**

- Installed K3s: curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=v1.22.16+k3s1 K3S_KUBECONFIG_MODE="644" INSTALL_K3S_EXEC="--flannel-backend=none --disable-network-policy --kube-apiserver-arg feature-gates=SeccompDefault=true" sh  -

**Expected behavior:**

 1. I enabled SeccompDefault feature when installing the cluster
 2. I created a simple nginx pod
 3. I checked the seccomp profile of the container and it should be the seccompprofile of the container runtime.
 

**Actual behavior:**

1. I enabled SeccompDefault feature when installing the cluster
 2. I created a simple nginx pod
 3. I checked the seccomp profile of the container and it's null : 
`sre@node-10-120-127-170:~$ sudo crictl inspect $(sudo crictl ps -q --name=nginx) | jq .info.runtimeSpec.linux.seccomp
null
sre@node-10-120-127-170:~$`

**Additional context / logs:**

**Comment by brandond** [comment]
You need to add the feature-gate to ALL the Kubernetes components that use it, not just the apiserver. This usually means the kubelet, controller-manager, and (occasionally) scheduler.

**Reply by brandond** [reply]
I will also note that this is Beta (on by default) in 1.25, so you could just install a 1.25 release of K3s if you want to play with it, rather than fiddling about with setting component args.

**Reply by jayzziebone** [reply]
I know that feature is Beta on 1.25 but the team is not ready to use 1.25 yet. So if I understand correctly I need, for each of these: --kube-scheduler-arg, --kube-controller-manager-arg and --kubelet-arg, to enable feature-gate. 
e.g: 
`--kube-scheduler-arg feature-gates=SeccompDefault=true, 
--kube-controller-manager-arg feature-gates=SeccompDefault=true, 
--kubelet-arg feature-gates=SeccompDefault=true`

Is this correct?

**Reply by brandond** [reply]
Yep, same arg for all of the components.

I usually like to run it all together as `--kubelet-arg=feature-gates=SeccompDefault=true` but your way should work fine too.

**Reply by jayzziebone** [reply]
Do I also need to do some changes on the agent side? Or just the server it's ok.

**Reply by jayzziebone** [reply]
I tried again and it's not working as it should. I kept it simple a ran a single node cluster.

1. Installed k3s:
`curl -sfL https://get.k3s.io | INSTALL_K3S_VERSION=v1.22.16+k3s1 K3S_KUBECONFIG_MODE="644" INSTALL_K3S_EXEC="--flannel-backend=none --disable-network-policy --kube-apiserver-arg=feature-gates=SeccompDefault=true --kubelet-arg=feature-gates=SeccompDefault=true --kube-controller-manager-arg=feature-gates=SeccompDefault=true --kube-scheduler-arg=feature-gates=SeccompDefault=true" sh -
`

2. This is the content of /etc/systemd/system/k3s.service:

[Unit]
Description=Lightweight Kubernetes
Documentation=https://k3s.io
Wants=network-online.target
After=network-online.target

[Install]
WantedBy=multi-user.target

[Service]
Type=notify
EnvironmentFile=-/etc/default/%N
EnvironmentFile=-/etc/sysconfig/%N
EnvironmentFile=-/etc/systemd/system/k3s.service.env
KillMode=process
Delegate=yes
`#Having non-zero Limit*s causes performance problems due to accounting overhead`
`# in the kernel. We recommend using cgroups to do container-local accounting.`
LimitNOFILE=1048576
LimitNPROC=infinity
LimitCORE=infinity
TasksMax=infinity
TimeoutStartSec=0
Restart=always
RestartSec=5s
ExecStartPre=/bin/sh -xc '! /usr/bin/systemctl is-enabled --quiet nm-cloud-setup.service'
ExecStartPre=-/sbin/modprobe br_netfilter
ExecStartPre=-/sbin/modprobe overlay
ExecStart=/usr/local/bin/k3s \
    server \
        '--flannel-backend=none' \
        '--disable-network-policy' \
        '--kube-apiserver-arg=feature-gates=SeccompDefault=true' \
        '--kubelet-arg=feature-gates=SeccompDefault=true' \
        '--kube-controller-manager-arg=feature-gates=SeccompDefault=true' \
        '--kube-scheduler-arg=feature-gates=SeccompDefault=true' \

3. I created a pod without `spec.securityContext.seccompProfile` in the pod descriptionand looked into the seccomp spec and still says null while it should pickup the profile from the container Runtime.

`
sre@node-10-120-13-171:~$ sudo crictl inspect $(sudo crictl ps -q --name=test-container) | jq .info.runtimeSpec.linux.seccomp
null
sre@node-10-120-13-171:~$
`

**Reply by brandond** [reply]
I am just reading the upstream docs at https://kubernetes.io/docs/tutorials/security/seccomp/#enable-the-use-of-runtimedefault-as-the-default-seccomp-profile-for-all-workloads and it says:
> To use seccomp profile defaulting, you must run the kubelet with the SeccompDefault [feature gate](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) enabled (this is the default). You must also explicitly enable the defaulting behavior for each node where you want to use this with the corresponding --seccomp-default [command line flag](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet). Both have to be enabled simultaneously to use the feature.

I don't see that you've added that flag to the kubelet?

**Reply by jayzziebone** [reply]
I added the flag to the kubelet too: --kubelet-arg=feature-gates=SeccompDefault=true', maybe something else is missing.

**Reply by brandond** [reply]
No, there's another CLI flag called`--seccomp-default` (a flag, not a feature-gate) that needs to be passed to the kubelet.

**Reply by jayzziebone** [reply]
I saw how to do it with k8s, but with k3s I'm not sure where to add this since k3s doesn't have a dedicated kubelet service. Do we have to create a /etc/rancher/k3s/config.yaml and add the flag in there?

Does this file also need a line like : apiVersion: kubelet.config.k8s.io/v1beta1 to start?

**Reply by brandond** [reply]
You're already setting kubelet args to set the feature-gate. You just need to add another --kubelet-arg for this flag.

I do note that you seem to have forgotten to escape the newlines on the ExecStart entry in your systemd unit, which probably isn't helping. If you're going to split the k3s args across multiple lines in the unit, it should look like this:

```
ExecStart=/usr/local/bin/k3s \
  server \
  '--flannel-backend=none' \
  '--disable-network-policy' \
  '--kubelet-arg=seccomp-default=true' \
  '--kubelet-arg=feature-gates=SeccompDefault=true' \
  '--kube-apiserver-arg=feature-gates=SeccompDefault=true' \
  '--kube-controller-manager-arg=feature-gates=SeccompDefault=true' \
  '--kube-scheduler-arg=feature-gates=SeccompDefault=true'
```

**Reply by jayzziebone** [reply]
Got it. Awesome! Thank you very much. Now it works. So it depends on every feature we want to enable, those that also require kubelet flag need extra arguments.
